libsolidity/ast/Types.cpp

```diff
@@ -2147,25 +2147,70 @@ TypePointer StructType::interfaceType(bool _inLibrary) const
 
 bool StructType::canBeUsedExternally(bool _inLibrary) const
 {
-	if (recursive())
-		return false;
-	else
+	if (_inLibrary && m_canBeUsedExternally_library.is_initialized())
+		return *m_canBeUsedExternally_library;
+
+	if (!_inLibrary && m_canBeUsedExternally.is_initialized())
+		return *m_canBeUsedExternally;
+
+	bool result = true;
+
+	auto visitor = [&](
+		StructDefinition const& _struct,
+		CycleDetector<StructDefinition>& _cycleDetector,
+		size_t /*_depth*/
+	)
 	{
 		// Check that all members have interface types.
-		// return false if at least one struct member does not have a type.
+		// Return false if at least one struct member does not have a type.
 		// This might happen, for example, if the type of the member does not exist,
 		// which is reported as an error.
-		for (auto const& var: m_struct.members())
+		for (ASTPointer<VariableDeclaration> const& variable: _struct.members())
 		{
 			// If the struct member does not have a type return false.
 			// A TypeError is expected in this case.
-			if (!var->annotation().type)
-				return false;
-			if (!var->annotation().type->canBeUsedExternally(_inLibrary))
-				return false;
+			if (!variable->annotation().type)
+			{
+				result = false;
+				return;
+			}
+
+			Type const* memberType = variable->annotation().type.get();
+
+			while (dynamic_cast<ArrayType const*>(memberType))
+				memberType = dynamic_cast<ArrayType const*>(memberType)->baseType().get();
+
+			if (StructType const* innerStruct = dynamic_cast<StructType const*>(memberType))
+				if (_cycleDetector.run(innerStruct->structDefinition()))
+				{
+					if (_inLibrary)
+						continue;
+					else
+					{
+						result = false;
+						return;
+					}
+				}
+
+			if (!memberType->canBeUsedExternally(_inLibrary))
```

Comment:
This line is reached for non-recursive structs - is that the intention?

Reply:
Yes, for "normal" (non-recursive) structs we still have to check each member whether it can be used externally

Reply:
But this will again run the cycle detector (a different instance, though) that we just ran on that struct, right? Even worse: What happens if `_inLibrary` is true - won't this lead to an infinite loop?

Reply:
This line is only reached if it's not a struct (or array of structs) that somewhere contains the current struct due to the check in line 2184. If the `if` gets a result it will either skip the member and look at the next (`_inLibrary => continue`) or stop the loop altogether (`return;`)

Reply:
So, it might run a cycle detector again if this is a struct, but at that point it can not be a struct that contains this struct.

Reply:
Ah sorry, I somehow did not recognize that the `continue` reaches that far outside :)